src/Microsoft.VisualStudio.Threading/AsyncMutex.cs

```diff
@@ -0,0 +1,193 @@
+    /// <exception cref="InvalidOperationException">Thrown if called before a prior call to this method has completed, with its releaser disposed if the mutex was entered.</exception>
+    public async ValueTask<LockReleaser> EnterAsync(TimeSpan timeout) => await this.TryEnterAsync(timeout) ?? throw new TimeoutException();
+
+    /// <summary>
+    /// Acquires the mutex asynchronously, allowing for timeouts without throwing exceptions.
+    /// </summary>
+    /// <param name="timeout">The maximum time to wait before timing out. Use <see cref="Timeout.InfiniteTimeSpan"/> for no timeout, or <see cref="TimeSpan.Zero"/> to acquire the mutex only if it is immediately available.</param>
+    /// <returns>
+    /// If the mutex was acquired, the result is a value whose disposal will release the mutex.
+    /// In the event of a timeout, the result in a <see langword="null" /> value.
+    /// </returns>
+    /// <exception cref="ArgumentOutOfRangeException">Thrown from the awaited result if the <paramref name="timeout"/> is a negative number other than -1 milliseconds, which represents an infinite timeout.</exception>
+    /// <exception cref="InvalidOperationException">Thrown if called before a prior call to this method has completed, with its releaser disposed if the mutex was entered.</exception>
+    public ValueTask<LockReleaser?> TryEnterAsync(TimeSpan timeout)
+    {
+        int priorState = Interlocked.CompareExchange(ref this.state, STATE_HELD_OR_WAITING, STATE_READY);
+        switch (priorState)
+        {
+            case STATE_HELD_OR_WAITING:
+                throw new InvalidOperationException();
+            case STATE_DISPOSED:
+                throw new ObjectDisposedException(this.GetType().FullName);
+        }
+
+        TaskCompletionSource<LockReleaser?> tcs = new();
+        this.mutexWorkQueue.Add(delegate
+        {
+            try
+            {
+                if (this.mutex.WaitOne(timeout))
+                {
+                    tcs.SetResult(new LockReleaser(this));
+                }
+                else
+                {
+                    Assumes.True(Interlocked.CompareExchange(ref this.state, STATE_READY, STATE_HELD_OR_WAITING) == STATE_HELD_OR_WAITING);
+                    tcs.SetResult(null);
+                }
+            }
+            catch (AbandonedMutexException)
+            {
+                tcs.SetResult(new LockReleaser(this, abandoned: true));
+            }
+            catch (Exception ex)
+            {
+                Assumes.True(Interlocked.CompareExchange(ref this.state, STATE_READY, STATE_HELD_OR_WAITING) == STATE_HELD_OR_WAITING);
+                tcs.SetException(ex);
+            }
+        });
+
+        return new ValueTask<LockReleaser?>(tcs.Task);
+    }
+
+    private void Release()
+    {
+        Assumes.True(Interlocked.CompareExchange(ref this.state, STATE_READY, STATE_HELD_OR_WAITING) == STATE_HELD_OR_WAITING);
+        this.mutexWorkQueue.Add(this.mutex.ReleaseMutex);
```

Comment:
in rare case, this may throw when the mutex is disposed, so it will throw InvalidOperationException from calling Dispose. not sure it is expected.

Reply:
If folks are disposing their mutex before releasing it, it's abandoning the mutex which is a serious problem in their code. If we throw an IOE at them when releasing in that case, that may help them find it. So I'm inclined to leave this one as-is.